perf/uncore_libpfm.go

```diff
@@ -0,0 +1,369 @@
+					return err
+				}
+				typeString := strings.TrimSpace(string(buf))
+				eventType, err := strconv.ParseUint(typeString, 0, 32)
+				if err != nil {
+					return err
+				}
+
+				buf, err = ioutil.ReadFile(filepath.Join(path, pmuCpumaskFilename))
+				if err != nil {
+					return err
+				}
+				var cpus []uint32
+				cpumask := strings.TrimSpace(string(buf))
+				for _, cpu := range cpumaskRegexp.Split(cpumask, -1) {
+					parsedCPU, err := strconv.ParseUint(cpu, 0, 32)
+					if err != nil {
+						return err
+					}
+					cpus = append(cpus, uint32(parsedCPU))
+				}
+
+				pmus = append(pmus, pmu{name: info.Name(), typeOf: uint32(eventType), cpus: cpus})
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pmus, nil
+}
+
+type uncoreCollector struct {
+	cpuFiles           map[string]map[uint32]map[int]readerCloser
+	cpuFilesLock       sync.Mutex
+	events             UncoreEvents
+	eventToCustomEvent map[Event]*CustomEvent
+}
+
+func NewUncoreCollector(configPath string) UncoreCollector {
+	if configPath == "" {
+		return &NoopUncoreCollector{}
+	}
+
+	configFile, err := os.Open(configPath)
+	if err != nil {
+		formatedError := fmt.Errorf("unable to read configuration file %q: %q", configPath, err)
+		klog.Warningf("Perf uncore metrics will not be available: %s", formatedError)
+		return &NoopUncoreCollector{}
+	}
+
+	config, err := parseConfig(configFile)
+	if err != nil {
+		formatedError := fmt.Errorf("unable to read configuration file %q: %q", configPath, err)
+		klog.Warningf("Perf uncore metrics will not be available: %s", formatedError)
+		return &NoopUncoreCollector{}
+	}
+
+	if areGroupedEventsUsed(config) {
+		err = fmt.Errorf("event grouping is not supported you must modify config file at %s", configPath)
+		klog.Warningf("Perf uncore metrics will not be available: %s", err)
+		return &NoopUncoreCollector{}
+	}
+
+	collector := &uncoreCollector{}
+
+	err = collector.setup(config)
+	if err != nil {
+		formatedError := fmt.Errorf("unable to setup uncore perf event collector: %v", err)
+		klog.Warningf("Perf uncore metrics will not be available: %s", formatedError)
+		return &NoopUncoreCollector{}
+	}
+
+	return collector
+}
+
+func (c *uncoreCollector) setup(events Events) error {
+	var err error
+	imcPMUs, err = getUncoreIMCPMUs(systemDevicesPath)
+	if err != nil {
+		return err
+	}
+
+	c.cpuFiles = make(map[string]map[uint32]map[int]readerCloser)
+	c.events, c.eventToCustomEvent = parseUncoreEvents(events)
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+
+	for _, group := range c.events.MemoryController {
+		customEvent, ok := c.eventToCustomEvent[group[0]]
+		var err error
+		if ok {
+			err = c.setupRawNonGroupedUncoreMemoryController(customEvent)
+		} else {
+			err = c.setupNonGroupedUncoreMemoryController(string(group[0]))
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func parseUncoreEvents(events Events) (UncoreEvents, map[Event]*CustomEvent) {
```

Comment:
Take a look at `perf.mapEventsToCustomEvents`. I think that this function can be simplified.

Reply:
Yes, it can be simplified but there'll be also info about core events.

Reply:
It makes me think that `perf.mapEventsToCustomEvents` should be modified not to include `UncoreEvents` then.